File: fontend/lib/principal.ts
/**
 * Principal console API client.
 *
 * This is the sole frontend data boundary for C-PR-01 … C-PR-10.  Components
 * consume these typed, tenant-scoped responses; no Principal screen imports an
 * in-memory data source or derives an institution metric in the browser.
 */

import { API_BASE_URL, getAccessToken, refreshAccessToken } from "./auth";
import { APIError, errorMessage, requestJson, guardTenantRefresh } from "./api-client";

const API_PREFIX = "principal";

export { APIError as PrincipalAPIError };

/** Shared tenant-console transport used by Principal and Vice Principal APIs. */
export function leadershipCall<T>(
  apiPrefix: string,
  path: string,
  init: RequestInit = {},
  errorName = "LeadershipAPIError",
): Promise<T> {
  return requestJson<T>(
    `${API_BASE_URL}/api/v1/${apiPrefix}${path}`,
    init,
    getAccessToken(),
    errorName,
    refreshAccessToken,
    guardTenantRefresh,
  );
}

const call = <T>(path: string, init: RequestInit = {}): Promise<T> =>
  leadershipCall<T>(API_PREFIX, path, init, "PrincipalAPIError");

export function queryString(values: Record<string, string | number | boolean | undefined | null>): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (value !== undefined && value !== null && value !== "") params.set(key, String(value));
  }
  const query = params.toString();
  return query ? `?${query}` : "";
}

// ── Shared shapes ───────────────────────────────────────────────────────────

export interface PrincipalPage<T> {
  total: number;
  limit: number;
  offset: number;
  items: T[];
}

export interface AttendanceClassSummary {
  id: string;
  name: string;
  attendance_percentage: number | null;
  total_present: number;
  total_absent: number;
  attendance_marks: number;
}

export interface AttendanceDepartmentSummary {
  id: string;
  name: string;
  attendance_percentage: number | null;
  total_present: number;
  total_absent: number;
  attendance_marks: number;
  classes: AttendanceClassSummary[];
}

export interface PrincipalAttendanceOverview {
  from_date: string;
  to_date: string;
  attendance_percentage: number | null;
  total_present: number;
  total_absent: number;
  attendance_marks: number;
  departments: AttendanceDepartmentSummary[];
}

export interface PrincipalUpcomingExam {
  id: string;
  title: string;
  scheduled_at: string;
  class_name: string;
  subject_name: string;
  department_name: string | null;
  status: string;
}

export interface PrincipalDashboard {
  academic_year: string | null;
  attendance_percentage: number | null;
  attendance_marks: number;
  attendance_departments: AttendanceDepartmentSummary[];
  ongoing_exams: number;
  upcoming_exams: number;
  upcoming_exam_items: PrincipalUpcomingExam[];
  pending_result_approvals: number;
  result_pass_percentage: number | null;
  staff_on_leave_today: number;
  staff_count: number;
  total_notices: number;
}

export interface PrincipalExamRow {
  id: string;
  title: string;
  class_id: string;
  class_name: string;
  department_name: string | null;
  subject_id: string;
  subject_name: string;
  subject_code: string;
  scheduled_at: string;
  window_end_at: string | null;
  duration_minutes: number;
  total_marks: number;
  passing_marks: number;
  mode: string;
  status: string;
  schedule_approval_status: "PENDING" | "APPROVED" | "REJECTED";
  schedule_approved_at: string | null;
  schedule_approval_note: string | null;
}

export interface PrincipalResultGroup {
  id: string;
  name: string;
  student_count: number;
  pass_count: number;
  fail_count: number;
  withheld_count: number;
  absent_count: number;
  pass_percentage: number | null;
  average_percentage: number | null;
}

export interface PrincipalPublicationRow {
  id: string;
  title: string;
  academic_year: string | null;
  class_name: string | null;
  published_at: string;
  published_by_name: string | null;
  exam_count: number;
  student_count: number;
  pass_percentage: number | null;
  average_percentage: number | null;
  is_visible_to_students: boolean;
  approval_status: "PENDING" | "APPROVED" | "REJECTED";
  approved_at: string | null;
  approval_note: string | null;
}

export interface PrincipalResultsOverview {
  overall: PrincipalResultGroup | null;
  departments: PrincipalResultGroup[];
  classes: PrincipalResultGroup[];
  publications: PrincipalPublicationRow[];
}

export interface PrincipalStaffRow {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  avatar_url: string | null;
  employee_code: string | null;
  designation: string | null;
  department_id: string | null;
  department_name: string | null;
  employment_type: string | null;
  date_of_joining: string | null;
  roles: string[];
  is_active: boolean;
}

export interface PrincipalStaffDetail extends PrincipalStaffRow {
  qualification: string | null;
  experience_years: number | null;
}

export interface PrincipalStudentEnrollment {
  class_id: string | null;
  class_name: string | null;
  department_name: string | null;
  academic_year_name: string | null;
  roll_number: string | null;
  status: string | null;
  enrollment_date: string | null;
}

export interface PrincipalStudentRow {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  avatar_url: string | null;
  roll_no: string | null;
  is_active: boolean;
  enrollment: PrincipalStudentEnrollment | null;
}

export interface PrincipalStudentDetail extends PrincipalStudentRow {
  date_of_birth: string | null;
  gender: string | null;
}

export interface PrincipalNoticeRow {
  id: string;
  title: string;
  body: string;
  author_name: string | null;
  target_scope: "INSTITUTION" | "DEPARTMENT" | "CLASS" | "HOSTEL" | "TRANSPORT";
  target_id: string | null;
  target_name: string | null;
  priority: "NORMAL" | "IMPORTANT" | "URGENT";
  is_pinned: boolean;
  published_at: string;
  expires_at: string | null;
  read_count: number;
  attachments: NoticeAttachment[];
}

export interface NoticeAttachment {
  id: string;
  file_name: string;
  file_size_bytes: number;
  mime_type: string;
  url: string;
  is_image: boolean;
  is_link: boolean;
}

export interface NoticeReader {
  id: string;
  name: string;
  read_at: string;
}

export interface PrincipalNoticeDetail extends PrincipalNoticeRow {
  readers: NoticeReader[];
}

export interface PrincipalTargetOption {
  id: string;
  name: string;
  department_id: string | null;
  department_name: string | null;
}

export interface PrincipalNoticeTargets {
  departments: PrincipalTargetOption[];
  classes: PrincipalTargetOption[];
}

export interface PrincipalTimetableSlot {
  id: string;
  class_id: string;
  class_name: string;
  department_name: string | null;
  day_of_week: number;
  period_number: number;
  start_time: string;
  end_time: string;
  subject_name: string | null;
  subject_code: string | null;
  teacher_name: string | null;
  room_no: string | null;
  slot_type: string;
}

export interface PrincipalTimetable {
  classes: PrincipalTargetOption[];
  slots: PrincipalTimetableSlot[];
}

export interface PrincipalPerformanceRow {
  department_id: string;
  department_name: string;
  attendance_percentage: number | null;
  pass_percentage: number | null;
  average_percentage: number | null;
  student_count: number;
}

export interface PrincipalReports {
  attendance: PrincipalAttendanceOverview;
  results: PrincipalResultsOverview;
  performance: PrincipalPerformanceRow[];
}

// ── Reads ───────────────────────────────────────────────────────────────────

export const fetchPrincipalDashboard = () => call<PrincipalDashboard>("/dashboard");

export const fetchPrincipalAttendance = (filters: { fromDate?: string; toDate?: string } = {}) =>
  call<PrincipalAttendanceOverview>(
    `/attendance${queryString({ from_date: filters.fromDate, to_date: filters.toDate })}`,
  );

export const fetchPrincipalExaminations = (
  filters: {
    status?: string;
    approvalStatus?: string;
    fromDate?: string;
    toDate?: string;
    limit?: number;
    offset?: number;
  } = {},
) =>
  call<PrincipalPage<PrincipalExamRow>>(
    `/examinations${queryString({
      status: filters.status,
      approval_status: filters.approvalStatus,
      from_date: filters.fromDate,
      to_date: filters.toDate,
      limit: filters.limit,
      offset: filters.offset,
    })}`,
  );

export const fetchPrincipalResults = () => call<PrincipalResultsOverview>("/results");

export const fetchPrincipalStaff = (
  filters: { query?: string; departmentId?: string; limit?: number; offset?: number } = {},
) =>
  call<PrincipalPage<PrincipalStaffRow>>(
    `/staff${queryString({
      query: filters.query,
      department_id: filters.departmentId,
      limit: filters.limit,
      offset: filters.offset,
    })}`,
  );

export const fetchPrincipalStaffDetail = (id: string) => call<PrincipalStaffDetail>(`/staff/${id}`);

export const fetchPrincipalStudents = (
  filters: { query?: string; classId?: string; limit?: number; offset?: number } = {},
) =>
  call<PrincipalPage<PrincipalStudentRow>>(
    `/students${queryString({
      query: filters.query,
      class_id: filters.classId,
      limit: filters.limit,
      offset: filters.offset,
    })}`,
  );

export const fetchPrincipalStudentDetail = (id: string) =>
  call<PrincipalStudentDetail>(`/students/${id}`);

export const fetchPrincipalNotices = (
  filters: {
    query?: string;
    scope?: "INSTITUTION" | "DEPARTMENT" | "CLASS";
    includeExpired?: boolean;
    limit?: number;
    offset?: number;
  } = {},
) =>
  call<PrincipalPage<PrincipalNoticeRow>>(
    `/notices${queryString({
      query: filters.query,
      scope: filters.scope,
      include_expired: filters.includeExpired,
      limit: filters.limit,
      offset: filters.offset,
    })}`,
  );

export const fetchPrincipalNotice = (id: string) => call<PrincipalNoticeDetail>(`/notices/${id}`);
export const fetchPrincipalNoticeTargets = () => call<PrincipalNoticeTargets>("/notices/targets");
export const fetchPrincipalTimetable = (classId?: string) =>
  call<PrincipalTimetable>(`/timetable${queryString({ class_id: classId })}`);
export const fetchPrincipalReports = (filters: { fromDate?: string; toDate?: string } = {}) =>
  call<PrincipalReports>(
    `/reports${queryString({ from_date: filters.fromDate, to_date: filters.toDate })}`,
  );

// ── Decisions / mutations ───────────────────────────────────────────────────

export const decideExamSchedule = (
  id: string,
  decision: "APPROVE" | "REJECT",
  note?: string,
) =>
  call<PrincipalExamRow>(`/examinations/${id}/approval`, {
    method: "POST",
    body: JSON.stringify({ decision, note: note?.trim() || undefined }),
  });

export const decideResultPublication = (
  id: string,
  decision: "APPROVE" | "REJECT",
  note?: string,
) =>
  call<PrincipalPublicationRow>(`/results/publications/${id}/approval`, {
    method: "POST",
    body: JSON.stringify({ decision, note: note?.trim() || undefined }),
  });

export const createPrincipalNotice = (payload: {
  title: string;
  body: string;
  target_scope: "INSTITUTION" | "DEPARTMENT" | "CLASS";
  target_id?: string | null;
  priority: "NORMAL" | "IMPORTANT" | "URGENT";
  is_pinned: boolean;
  expires_at?: string | null;
  attachments?: Array<{ file_name: string; mime_type: string; data_url?: string; external_url?: string }>;
}) =>
  call<PrincipalNoticeDetail>("/notices", {
    method: "POST",
    body: JSON.stringify(payload),
  });

// ── CSV export ──────────────────────────────────────────────────────────────

/**
 * Download a leadership aggregate report. The backend sends CSV rather than
 * an API envelope, so this shared transport keeps both leadership consoles
 * consistent without duplicating bearer/error/download handling.
 */
export async function downloadLeadershipCsv(
  apiPrefix: string,
  exportPath: string,
  filenameFallback: string,
  query: Record<string, string | number | boolean | undefined | null> = {},
): Promise<void> {
  const token = getAccessToken();
  const response = await fetch(
    `${API_BASE_URL}/api/v1/${apiPrefix}${exportPath}${queryString(query)}`,
    {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      credentials: "include",
    },
  );
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new APIError(errorMessage(body, response.status), response.status, "LeadershipAPIError");
  }

  const disposition = response.headers.get("Content-Disposition") ?? "";
  const name = /filename="?([^";]+)"?/i.exec(disposition)?.[1] ?? filenameFallback;
  const objectUrl = URL.createObjectURL(await response.blob());
  const anchor = document.createElement("a");
  anchor.href = objectUrl;
  anchor.download = name;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(objectUrl);
}

export function downloadLeadershipReport(
  apiPrefix: string,
  kind: string,
  filters: { fromDate?: string; toDate?: string } = {},
): Promise<void> {
  return downloadLeadershipCsv(apiPrefix, "/reports/export", `${kind}-report.csv`, {
    kind,
    from_date: filters.fromDate,
    to_date: filters.toDate,
  });
}

export function downloadPrincipalReport(
  kind: "attendance" | "results" | "performance" | "timetable" | "examinations",
  filters: { fromDate?: string; toDate?: string } = {},
): Promise<void> {
  return downloadLeadershipReport(API_PREFIX, kind, filters);
}

File: backend/app/schemas/principal.py
"""Wire contracts for the Principal console (C-PR-01 … C-PR-10).

The contracts deliberately contain aggregates and read-only directory data, not
raw operational rows.  A Principal needs institution-wide academic oversight;
they do not need payroll, passwords, bank data or another role's mutation
controls.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.common import APIResponse


# ── Shared paging / directory contracts ──────────────────────────────────────

class PrincipalPage(BaseModel):
    total: int
    limit: int
    offset: int


class PrincipalStaffRow(BaseModel):
    id: uuid.UUID
    name: str
    email: str | None = None
    phone: str | None = None
    avatar_url: str | None = None
    employee_code: str | None = None
    designation: str | None = None
    department_id: uuid.UUID | None = None
    department_name: str | None = None
    employment_type: str | None = None
    date_of_joining: date | None = None
    roles: list[str] = Field(default_factory=list)
    is_active: bool


class PrincipalStaffDetail(PrincipalStaffRow):
    qualification: str | None = None
    experience_years: int | None = None


class PrincipalStudentEnrollment(BaseModel):
    class_id: uuid.UUID | None = None
    class_name: str | None = None
    department_name: str | None = None
    academic_year_name: str | None = None
    roll_number: str | None = None
    status: str | None = None
    enrollment_date: date | None = None


class PrincipalStudentRow(BaseModel):
    id: uuid.UUID
    name: str
    email: str | None = None
    phone: str | None = None
    avatar_url: str | None = None
    roll_no: str | None = None
    is_active: bool
    enrollment: PrincipalStudentEnrollment | None = None


class PrincipalStudentDetail(PrincipalStudentRow):
    date_of_birth: date | None = None
    gender: str | None = None


class PrincipalStaffPage(PrincipalPage):
    items: list[PrincipalStaffRow]


class PrincipalStudentPage(PrincipalPage):
    items: list[PrincipalStudentRow]


# ── Dashboard / attendance ──────────────────────────────────────────────────

class AttendanceClassSummary(BaseModel):
    id: uuid.UUID
    name: str
    attendance_percentage: float | None = None
    total_present: int
    total_absent: int
    attendance_marks: int


class AttendanceDepartmentSummary(BaseModel):
    id: uuid.UUID
    name: str
    attendance_percentage: float | None = None
    total_present: int
    total_absent: int
    attendance_marks: int
    classes: list[AttendanceClassSummary] = Field(default_factory=list)


class PrincipalAttendanceOverview(BaseModel):
    from_date: date
    to_date: date
    attendance_percentage: float | None = None
    total_present: int
    total_absent: int
    attendance_marks: int
    departments: list[AttendanceDepartmentSummary]


class PrincipalUpcomingExam(BaseModel):
    id: uuid.UUID
    title: str
    scheduled_at: datetime
    class_name: str
    subject_name: str
    department_name: str | None = None
    status: str


class PrincipalDashboard(BaseModel):
    academic_year: str | None = None
    attendance_percentage: float | None = None
    attendance_marks: int
    attendance_departments: list[AttendanceDepartmentSummary]
    ongoing_exams: int
    upcoming_exams: int
    upcoming_exam_items: list[PrincipalUpcomingExam]
    pending_result_approvals: int
    result_pass_percentage: float | None = None
    staff_on_leave_today: int
    staff_count: int
    total_notices: int


# ── Exams / schedule approval ────────────────────────────────────────────────

class PrincipalExamRow(BaseModel):
    id: uuid.UUID
    title: str
    class_id: uuid.UUID
    class_name: str
    department_name: str | None = None
    subject_id: uuid.UUID
    subject_name: str
    subject_code: str
    scheduled_at: datetime
    window_end_at: datetime | None = None
    duration_minutes: int
    total_marks: int
    passing_marks: int
    mode: str
    status: str
    schedule_approval_status: Literal["PENDING", "APPROVED", "REJECTED"]
    schedule_approved_at: datetime | None = None
    schedule_approval_note: str | None = None


class PrincipalExamPage(PrincipalPage):
    items: list[PrincipalExamRow]


class ScheduleApprovalRequest(BaseModel):
    decision: Literal["APPROVE", "REJECT"]
    note: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def rejection_needs_reason(self) -> "ScheduleApprovalRequest":
        if self.decision == "REJECT" and not (self.note or "").strip():
            raise ValueError("A reason is required when rejecting an exam schedule")
        return self


# ── Results / publication approval ──────────────────────────────────────────

class PrincipalResultGroup(BaseModel):
    id: uuid.UUID
    name: str
    student_count: int
    pass_count: int
    fail_count: int
    withheld_count: int
    absent_count: int
    pass_percentage: float | None = None
    average_percentage: float | None = None


class PrincipalPublicationRow(BaseModel):
    id: uuid.UUID
    title: str
    academic_year: str | None = None
    class_name: str | None = None
    published_at: datetime
    published_by_name: str | None = None
    exam_count: int
    student_count: int
    pass_percentage: float | None = None
    average_percentage: float | None = None
    is_visible_to_students: bool
    approval_status: Literal["PENDING", "APPROVED", "REJECTED"]
    approved_at: datetime | None = None
    approval_note: str | None = None


class PrincipalResultsOverview(BaseModel):
    overall: PrincipalResultGroup | None = None
    departments: list[PrincipalResultGroup]
    classes: list[PrincipalResultGroup]
    publications: list[PrincipalPublicationRow]


class ResultApprovalRequest(BaseModel):
    decision: Literal["APPROVE", "REJECT"]
    note: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def rejection_needs_reason(self) -> "ResultApprovalRequest":
        if self.decision == "REJECT" and not (self.note or "").strip():
            raise ValueError("A reason is required when rejecting a result publication")
        return self


# ── Notice board ─────────────────────────────────────────────────────────────

class NoticeAttachmentInput(BaseModel):
    """A browser-selected file (data URL) or a safe external link."""

    file_name: str = Field(..., min_length=1, max_length=255)
    mime_type: str = Field(..., min_length=1, max_length=100)
    data_url: str | None = None
    external_url: str | None = Field(default=None, max_length=2048)

    @model_validator(mode="after")
    def has_one_source(self) -> "NoticeAttachmentInput":
        if bool(self.data_url) == bool(self.external_url):
            raise ValueError("Provide either a file or a link attachment")
        if self.external_url and not self.external_url.startswith(("https://", "http://")):
            raise ValueError("Links must use http or https")
        return self


class NoticeAttachment(BaseModel):
    id: uuid.UUID
    file_name: str
    file_size_bytes: int
    mime_type: str
    url: str
    is_image: bool
    is_link: bool


class LeadershipNoticeRow(BaseModel):
    """Notice metadata shared by leadership roles without receipt data."""

    id: uuid.UUID
    title: str
    body: str
    author_name: str | None = None
    # Principal reads every institution notice, including optional-module
    # notices. Posting remains restricted to the three academic scopes below.
    target_scope: Literal["INSTITUTION", "DEPARTMENT", "CLASS", "HOSTEL", "TRANSPORT"]
    target_id: uuid.UUID | None = None
    target_name: str | None = None
    priority: Literal["NORMAL", "IMPORTANT", "URGENT"]
    is_pinned: bool
    published_at: datetime
    expires_at: datetime | None = None
    attachments: list[NoticeAttachment] = Field(default_factory=list)


class PrincipalNoticeRow(LeadershipNoticeRow):
    """Principal-only row adds the aggregate read receipt count."""

    read_count: int


class PrincipalNoticePage(PrincipalPage):
    items: list[PrincipalNoticeRow]


class NoticeReader(BaseModel):
    id: uuid.UUID
    name: str
    read_at: datetime


class PrincipalNoticeDetail(PrincipalNoticeRow):
    readers: list[NoticeReader]


class PrincipalNoticeCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=255)
    body: str = Field(..., min_length=1, max_length=20_000)
    target_scope: Literal["INSTITUTION", "DEPARTMENT", "CLASS"]
    target_id: uuid.UUID | None = None
    priority: Literal["NORMAL", "IMPORTANT", "URGENT"] = "NORMAL"
    is_pinned: bool = False
    expires_at: datetime | None = None
    attachments: list[NoticeAttachmentInput] = Field(default_factory=list, max_length=5)

    @field_validator("expires_at")
    @classmethod
    def expiry_must_include_timezone(cls, value: datetime | None) -> datetime | None:
        if value is not None and (value.tzinfo is None or value.utcoffset() is None):
            raise ValueError("expires_at must include a timezone")
        return value

    @model_validator(mode="after")
    def validate_target(self) -> "PrincipalNoticeCreate":
        if self.target_scope == "INSTITUTION" and self.target_id is not None:
            raise ValueError("Institution-wide notices cannot have a target id")
        if self.target_scope != "INSTITUTION" and self.target_id is None:
            raise ValueError("A department or class target is required")
        return self


class PrincipalTargetOption(BaseModel):
    id: uuid.UUID
    name: str
    department_id: uuid.UUID | None = None
    department_name: str | None = None


class PrincipalNoticeTargets(BaseModel):
    departments: list[PrincipalTargetOption]
    classes: list[PrincipalTargetOption]


# ── Timetable ────────────────────────────────────────────────────────────────

class PrincipalTimetableSlot(BaseModel):
    id: uuid.UUID
    class_id: uuid.UUID
    class_name: str
    department_name: str | None = None
    day_of_week: int = Field(..., ge=1, le=7)
    period_number: int
    start_time: time
    end_time: time
    subject_name: str | None = None
    subject_code: str | None = None
    teacher_name: str | None = None
    room_no: str | None = None
    slot_type: str


class PrincipalTimetable(BaseModel):
    classes: list[PrincipalTargetOption]
    slots: list[PrincipalTimetableSlot]


# ── Reports ──────────────────────────────────────────────────────────────────

class PrincipalPerformanceRow(BaseModel):
    department_id: uuid.UUID
    department_name: str
    attendance_percentage: float | None = None
    pass_percentage: float | None = None
    average_percentage: float | None = None
    student_count: int


class PrincipalReports(BaseModel):
    attendance: PrincipalAttendanceOverview
    results: PrincipalResultsOverview
    performance: list[PrincipalPerformanceRow]


# ── API envelope aliases ─────────────────────────────────────────────────────

APIResponsePrincipalDashboard = APIResponse[PrincipalDashboard]
APIResponsePrincipalAttendance = APIResponse[PrincipalAttendanceOverview]
APIResponsePrincipalExams = APIResponse[PrincipalExamPage]
APIResponsePrincipalExam = APIResponse[PrincipalExamRow]
APIResponsePrincipalResults = APIResponse[PrincipalResultsOverview]
APIResponsePrincipalPublication = APIResponse[PrincipalPublicationRow]
APIResponsePrincipalStaff = APIResponse[PrincipalStaffPage]
APIResponsePrincipalStaffDetail = APIResponse[PrincipalStaffDetail]
APIResponsePrincipalStudents = APIResponse[PrincipalStudentPage]
APIResponsePrincipalStudentDetail = APIResponse[PrincipalStudentDetail]
APIResponsePrincipalNotices = APIResponse[PrincipalNoticePage]
APIResponsePrincipalNotice = APIResponse[PrincipalNoticeDetail]
APIResponsePrincipalNoticeTargets = APIResponse[PrincipalNoticeTargets]
APIResponsePrincipalTimetable = APIResponse[PrincipalTimetable]
APIResponsePrincipalReports = APIResponse[PrincipalReports]
